Mathlib/Algebra/Order/Quantale.lean

```diff
@@ -0,0 +1,160 @@
+a frame if and only if it is integral and idempotent are developed separately in order to reduce
+overhead if a user does not need them.
+
+## Main definitions
+
+* class `Quantale`: a semigroup distributing over a complete lattice, i.e satisfying
+  `x * (sSup s) = ⨆ y ∈ s, x * y` and `(sSup s) * y = ⨆ x ∈ s, x * y`;
+
+* `IsIntegralQuantale`: Typeclass mixin for a unital quantale (i.e. a quantale in which the
+  semigroup is a monoid) respecting `⊤ = 1` (also called a two-sided quantale);
+
+* next to these classes, we define the additive versions `AddQuantale`, `IsAddIntegral` in
+  which the semigroup operation is denoted by addition instead of multiplication;
+
+## Naming conventions
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the leftResiduation of `y` over `x`;
+
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the rightResiduation of `y` over `x`;
+
+## References
+
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale>
+
+-/
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class AddQuantale (α : Type _) [AddSemigroup α] extends CompleteLattice α where
+  /-- Addition is distributive over join in a quantale -/
+  protected add_sSup_eq_iSup_add (x : α) (s : Set α) : x + sSup s = ⨆ y ∈ s, x + y
+  /-- Addition is distributive over join in a quantale -/
+  protected sSup_add_eq_iSup_add (s : Set α) (y : α) : sSup s + y = ⨆ x ∈ s, x + y
+
+/-- A quantale is a semigroup distributing over a complete lattice. -/
+@[to_additive]
+class Quantale (α : Type _) [Semigroup α] extends CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
+class IsIntegralAddQuantale (α : Type _) [AddMonoid α] [AddQuantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_zero : (⊤ : α) = 0
+
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegralQuantale (α : Type _) [Monoid α] [Quantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_one : (⊤ : α) = 1
+
+section Quantale
+
+variable (α : Type _)
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
```

Comment:
Can we have the `iSup` version?

Reply:
Not sure which version you are aiming at.
I'm just generalizing the definition on frames here
(see CompleteBooleanAlgebra).

class Order.Frame (α : Type*) extends CompleteLattice α, HeytingAlgebra α where
  /-- `⊓` distributes over `⨆`. -/
  inf_sSup_le_iSup_inf (a : α) (s : Set α) : a ⊓ sSup s ≤ ⨆ b ∈ s, a ⊓ b

Question is, perhaps, whether 'le' is enough or 'eq' is necessary.

Reply:
As mul is not commutative, we need two of course.

Reply:
I found the iSup version for Frames and added a number of theorems for Quantales based on what I found for Frames in CompleteBooleanAlgebra.

Now there are iSup and sup versions, as wel as monotonicity (mul_le_mul). Those help to prove the equivalence on residuals you asked for later as well.